pkg/controller/azvolumeattachment.go

```diff
@@ -0,0 +1,546 @@
+	// TODO: add detach volume function from provisioner library
+	return nil
+}
+
+func NewAzVolumeAttachmentController(ctx context.Context, mgr manager.Manager, azVolumeClient *azVolumeClientSet.Interface, namespace string) error {
+	c, err := controller.New("azvolumeattachment-controller", mgr, controller.Options{
+		MaxConcurrentReconciles: 10,
+		Reconciler:              &reconcileAzVolumeAttachment{client: mgr.GetClient(), azVolumeClient: *azVolumeClient, namespace: namespace},
+		Log:                     mgr.GetLogger().WithValues("controller", "azdrivernode"),
+	})
+
+	if err != nil {
+		klog.Errorf("failed to create a new azvolumeattachment controller: (%v)", err)
+		return err
+	}
+
+	// Watch for CRUD events on azVolumeAttachment objects
+	err = c.Watch(&source.Kind{Type: &v1alpha1.AzVolumeAttachment{}}, &handler.Funcs{
+		CreateFunc: func(e event.CreateEvent, q workqueue.RateLimitingInterface) {
+			if e.Object == nil {
+				klog.Warning("received empty object")
+				return
+			}
+			q.Add(reconcile.Request{NamespacedName: types.NamespacedName{
+				Name:      e.Object.GetName(),
+				Namespace: e.Object.GetNamespace(),
+			}})
+		},
+		UpdateFunc: func(e event.UpdateEvent, q workqueue.RateLimitingInterface) {
+			if e.ObjectNew == nil || e.ObjectOld == nil {
+				return
+			}
+			// if it is an update in state from AttachRequired to Attached, ignore
+			go func() {
+				var oldAttachment, newAttachment v1alpha1.AzVolumeAttachment
+				err1 := mgr.GetClient().Get(ctx, types.NamespacedName{Name: e.ObjectOld.GetName(), Namespace: e.ObjectOld.GetNamespace()}, &oldAttachment)
+				err2 := mgr.GetClient().Get(ctx, types.NamespacedName{Name: e.ObjectNew.GetName(), Namespace: e.ObjectNew.GetNamespace()}, &newAttachment)
+				if err1 != nil || err2 != nil {
+					return
+				}
+				if oldAttachment.Status.AttachmentState == v1alpha1.AttachRequired && newAttachment.Status.AttachmentState == v1alpha1.Attached {
+					return
+				}
+				q.Add(reconcile.Request{NamespacedName: types.NamespacedName{
+					Name:      e.ObjectNew.GetName(),
+					Namespace: e.ObjectNew.GetNamespace(),
+				}})
+			}()
+		}})
+
+	if err != nil {
+		klog.Errorf("failed to initialize watch for azvolumeattachment object: (%v)", err)
+		return err
+	}
+
+	// Watch for Update events on storage class objects (managed-csi-shared-with-replica explicitly)
+	err = c.Watch(&source.Kind{Type: &storagev1.StorageClass{}}, &handler.Funcs{
+		UpdateFunc: func(e event.UpdateEvent, q workqueue.RateLimitingInterface) {
+			// only add to the queue if the object's provisioner is disk.csi.azure.com and has MaxShares parameter
+			go func() {
+				var storageClass storagev1.StorageClass
+				if err := mgr.GetClient().Get(ctx, types.NamespacedName{Name: e.ObjectNew.GetName(), Namespace: e.ObjectNew.GetNamespace()}, &storageClass); err != nil {
+					return
+				} else if _, ok := storageClass.Parameters["maxShares"]; storageClass.Provisioner == DriverName && ok {
```

Comment:
It would be good to understand the scenarios where a change to the sharing parameters would be desirable. The `StorageClass` is intended to advertise the capabilities of the platform, and not really as a way to make global configuration changes. At the end of the day, how likely is it for a customer to change the `StorageClass` in a production scenario that would have a benefit? If a customer needs to change the storage for a pod, the expectation is for them to create a new `StorageClass`

Reply:
So does that mean there is no option for the users dynamically change number of replicas they want for their volumes running on pods?

Reply:
There are multiple interconnected things, if we are allowing to change max replica count then we should allow to change maxShares and today maxShares is set at the time of PV creation only.

So I think for simplicity and consistency, we should not allow changes to this.

@sunpa93 brings up a very interesting point as to what if customers want to increase the replica count. My initial though process here was that we let customers set the maxShares and an indication that they want replicas and we allow the driver to adjust the number of attachments in the backend in the managed fashion. What that means is:

1. If maxShares is 2, then we only have one replica attachment.
2. We go up to 2 replica attachments and no of replica should always be <= maxShares - 1.
3. Volume attachment control should be able to increase\decrease the replica attachment count based on disk pressure on the cluster.

what this does is:
1. Provides the back-up attachment, without having users to worry about maxShares vs max replica attachment confusion.
2. Optimize the number of attachments for them, since there is a cost associated with each attachment, they can accidently set it to a very high value.
3. Avoid customers to make mistakes like setting max replica count to a very high value and then causing disk pressure which could lead to pod placement issues.
4. Give flexibility to the driver to increase\decrease replicas to provide high availability and manage disk pressure\cost.

Let me know what you folks think.